### PR TITLE
v.univar: add JSON support

### PR DESCRIPTION
**Pull request by kritibirda26**
Use parson to add json output format support to the v.univar module.

Expected JSON schema:
Root is a JSON object. The percentile option allows the user to request specific percentile which is written as a percentile_%d key to JSON object.

```json
{
    "n": <int>,
    "missing": <int>,
    "nnull": <int>,
    "min": <double>,
    "max": <double>,
    "range": <double>,
    "sum": <double>,
    "mean": <double>,
    "mean_abs": <double>,
    "population_stddev": <double>,
    "population_variance": <double>,
    "population_coeff_variation": <double>,
    "sample_stddev": <double>,
    "sample_variance": <double>,
    "kurtosis": <double>,
    "skewness": <double>,
    "first_quartile": <double>,
    "median": <double>,
    "third_quartile": <double>,
    "percentile_90": <double>
}
```

### CODE REVIEW

**Comment by echoix** [comment]
Special naming of keys would be hard to use from another tool. It might be better to have a list, but since we need the value and the key, it would need a mapping, so something like that? : 

```json
{
  ...
  "percentile": { 
    "5":  0.536736,
    "90": 0.7237272,
    "98": 0.863662721,
    "99.5": 0.916363
   },
  ...
}
```

**Comment by echoix** [comment]
A good exercise to know if the json output is really usable from a tool (to be able to be used in an automated way), is to try to create a real jsonschema. If it's too hard to describe the format with a json schema, then it probably isn't usable, and can't be validated against that schema.

**Comment by echoix** [comment]
I tried with an online tool, https://jsonschema.net/app/schemas/456991 to see what it looks like (by generating the skeleton of a jsonschema). I had to change your example to be a valid json first, I took:

```json

{
    "n": 55,
    "missing": 55,
    "nnull": 55,
    "min": 1.234,
    "max": 1.2345,
    "range": 1.23456,
    "sum": 1.234569,
    "mean": 1.234567,
    "mean_abs": 1.234568,
    "population_stddev": 2.3456,
    "population_variance": 2.34567,
    "population_coeff_variation": 2.345678,
    "sample_stddev": 4.5556,
    "sample_variance": 6.7777,
    "kurtosis": 3.4455,
    "skewness": 6.8888,
    "first_quartile": 8.8888,
    "median": 7.777,
    "third_quartile": 8.8882,
    "percentile_90": 6.7776
}
```

Gives the jsonschema:
```json
{
    "$schema": "https://json-schema.org/draft/2019-09/schema",
    "$id": "http://example.com/example.json",
    "type": "object",
    "default": {},
    "title": "Root Schema",
    "required": [
        "n",
        "missing",
        "nnull",
        "min",
        "max",
        "range",
        "sum",
        "mean",
        "mean_abs",
        "population_stddev",
        "population_variance",
        "population_coeff_variation",
        "sample_stddev",
        "sample_variance",
        "kurtosis",
        "skewness",
        "first_quartile",
        "median",
        "third_quartile",
        "percentile_90"
    ],
    "properties": {
        "n": {
            "type": "integer",
            "default": 0,
            "title": "The n Schema",
            "examples": [
                55
            ]
        },
        "missing": {
            "type": "integer",
            "default": 0,
            "title": "The missing Schema",
            "examples": [
                55
            ]
        },
        "nnull": {
            "type": "integer",
            "default": 0,
            "title": "The nnull Schema",
            "examples": [
                55
            ]
        },
        "min": {
            "type": "number",
            "default": 0.0,
            "title": "The min Schema",
            "examples": [
                1.234
            ]
        },
        "max": {
            "type": "number",
            "default": 0.0,
            "title": "The max Schema",
            "examples": [
                1.2345
            ]
        },
        "range": {
            "type": "number",
            "default": 0.0,
            "title": "The range Schema",
            "examples": [
                1.23456
            ]
        },
        "sum": {
            "type": "number",
            "default": 0.0,
            "title": "The sum Schema",
            "examples": [
                1.234569
            ]
        },
        "mean": {
            "type": "number",
            "default": 0.0,
            "title": "The mean Schema",
            "examples": [
                1.234567
            ]
        },
        "mean_abs": {
            "type": "number",
            "default": 0.0,
            "title": "The mean_abs Schema",
            "examples": [
                1.234568
            ]
        },
        "population_stddev": {
            "type": "number",
            "default": 0.0,
            "title": "The population_stddev Schema",
            "examples": [
                2.3456
            ]
        },
        "population_variance": {
            "type": "number",
            "default": 0.0,
            "title": "The population_variance Schema",
            "examples": [
                2.34567
            ]
        },
        "population_coeff_variation": {
            "type": "number",
            "default": 0.0,
            "title": "The population_coeff_variation Schema",
            "examples": [
                2.345678
            ]
        },
        "sample_stddev": {
            "type": "number",
            "default": 0.0,
            "title": "The sample_stddev Schema",
            "examples": [
                4.5556
            ]
        },
        "sample_variance": {
            "type": "number",
            "default": 0.0,
            "title": "The sample_variance Schema",
            "examples": [
                6.7777
            ]
        },
        "kurtosis": {
            "type": "number",
            "default": 0.0,
            "title": "The kurtosis Schema",
            "examples": [
                3.4455
            ]
        },
        "skewness": {
            "type": "number",
            "default": 0.0,
            "title": "The skewness Schema",
            "examples": [
                6.8888
            ]
        },
        "first_quartile": {
            "type": "number",
            "default": 0.0,
            "title": "The first_quartile Schema",
            "examples": [
                8.8888
            ]
        },
        "median": {
            "type": "number",
            "default": 0.0,
            "title": "The median Schema",
            "examples": [
                7.777
            ]
        },
        "third_quartile": {
            "type": "number",
            "default": 0.0,
            "title": "The third_quartile Schema",
            "examples": [
                8.8882
            ]
        },
        "percentile_90": {
            "type": "number",
            "default": 0.0,
            "title": "The percentile_90 Schema",
            "examples": [
                6.7776
            ]
        }
    },
    "examples": [{
        "n": 55,
        "missing": 55,
        "nnull": 55,
        "min": 1.234,
        "max": 1.2345,
        "range": 1.23456,
        "sum": 1.234569,
        "mean": 1.234567,
        "mean_abs": 1.234568,
        "population_stddev": 2.3456,
        "population_variance": 2.34567,
        "population_coeff_variation": 2.345678,
        "sample_stddev": 4.5556,
        "sample_variance": 6.7777,
        "kurtosis": 3.4455,
        "skewness": 6.8888,
        "first_quartile": 8.8888,
        "median": 7.777,
        "third_quartile": 8.8882,
        "percentile_90": 6.7776
    }]
}
```

**Comment by wenzeslaus** [comment]
> Special naming of keys would be hard to use from another tool. It might be better to have a list, but since we need the value and the key, it would need a mapping...

I was actually thinking I used two lists in db.univar, but it turns out I used whatever was in the shell scriptstyle format, so list of keys which depends on the input. It depends on the input, so you can figure the keys out. I think that was my logic there. It is terrible for writing a schema and that's a valid criticism. I think we can change than in a way which would be backwards compatible and make it consistent across all percentile outputs at the same time if we decide to go different direction.

**Comment by echoix** [comment]
> I was actually thinking I used two lists in db.univar, but it turns out I used whatever was in the shell scriptstyle format, so list of keys which depends on the input. It depends on the input, so you can figure the keys out. I think that was my logic there. It is terrible for writing a schema and that's a valid criticism. I think we can change than in a way which would be backwards compatible and make it consistent across all percentile outputs at the same time if we decide to go different direction.

Was the output of db.univar with the two lists released yet? If not, it's not a breaking change yet.

**Comment by wenzeslaus** [comment]
> Was the output of db.univar with the two lists released yet?

Yes, it was, in 8.3.0. https://github.com/OSGeo/grass/commit/120f198fca22a564334d7721fabbe8c6e299bf6b

I think the situation now is: 1) We should do whatever is right here not to spread a bad design from one tool to another. 2) We might be able to add the better percentile output to v.db.univar depending on the keys we choose here.

So, let's decide the percentiles here.

I guess the following is lengthy, but easy to express in a schema:

```json
{
"percentiles": [
    {"percentile": 95, "value": 200.03},
    {"percentile": 99.9, "value": 220.01}
]
}
```

**Comment by echoix** [comment]
> > Was the output of db.univar with the two lists released yet?
> 
> Yes, it was, in 8.3.0. [120f198](https://github.com/OSGeo/grass/commit/120f198fca22a564334d7721fabbe8c6e299bf6b)
> 
> I think the situation now is: 1) We should do whatever is right here not to spread a bad design from one tool to another. 2) We might be able to add the better percentile output to v.db.univar depending on the keys we choose here.
> 
> So, let's decide the percentiles here.
> 
> I guess the following is lengthy, but easy to express in a schema:
> 
> ```json
> {
> "percentiles": [
>     {"percentile": 95, "value": 200.03},
>     {"percentile": 99.9, "value": 220.01}
> ]
> }
> ```

It’s lengthy, yes, but I like the fact that a list is used here, as you can repeat the same percentile if needed, and it allows to keep the same order of percentiles requested in input.
I didn’t have an idea how to keep the ordering in my first example.

**Comment by wenzeslaus** [comment]
What do you this @kritibirda26, does the list of dictionaries look good to you?

**Comment by kritibirda26** [comment]
Hi @wenzeslaus and @echoix, The dictionaries for percentiles make sense to me. I'll update the format.

**Comment by cwhite911** [comment]
> Hi @wenzeslaus and @echoix, Sorry for the delay in the response. The dictionaries for percentiles make sense to me. I'll update the format.

Are you referring to the schema suggested by @echoix? 

> > ```json
> > {
> > "percentiles": [
> >     {"percentile": 95, "value": 200.03},
> >     {"percentile": 99.9, "value": 220.01}
> > ]
> > }
> > ```

This is the approach I think we should take.

File: vector/v.univar/main.c [hunk 1]
```diff
@@ -688,7 +748,16 @@ void summary(void)
...
         }
 
-        if (shell_flag->answer) {
+        if (format == JSON) {
+            json_object_set_number(root_object, "first_quartile", quartile_25);
+            json_object_set_number(root_object, "median", median);
+            json_object_set_number(root_object, "third_quartile", quartile_75);
+
+            char percentile_key[24];
+            snprintf(percentile_key, 24, "percentile_%d", perc);
```
Comment (cwhite911) [inline vector/v.univar/main.c:757]: Update to match the json schema from our discussion.

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "percentiles": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "percentile": {
            "type": "number",
            "minimum": 0,
            "maximum": 100
          },
          "value": {
            "type": "number"
          }
        },
        "required": ["percentile", "value"],
        "additionalProperties": false
      }
    }
  },
  "required": ["percentiles"],
  "additionalProperties": false
}

```

File: vector/v.univar/v.univar.html [hunk 2]
```diff
@@ -121,6 +121,36 @@ <h3>Calculate statistic of distance between sampling points</h3>
...
+    "population_variance": 322.20958540007473,
+    "population_coeff_variation": 0.165604714023907,
+    "sample_stddev": 18.04062714200488,
+    "sample_variance": 325.46422767684317,
+    "kurtosis": -0.80852318980579341,
+    "skewness": 0.24812587332362507,
+    "first_quartile": 94.531009999999995,
+    "median": 104.48525000000001,
+    "third_quartile": 121.6682,
+    "percentile_90": 133.7296
```
Comment (cwhite911) [inline vector/v.univar/v.univar.html:150]: Update docs to match the percentiles

**Review by cwhite911 — CHANGES_REQUESTED** [review]
Looking good. Please update the percentiles, tests, and docs.

**Comment by kritibirda26** [comment]
Yes, on it.

File: vector/v.univar/v.univar.html (hunk 2 above)
Comment (kritibirda26) [reply vector/v.univar/v.univar.html]: Done

File: vector/v.univar/main.c (hunk 1 above)
Comment (kritibirda26) [reply vector/v.univar/main.c]: Done.

**Review by cwhite911 — APPROVED** [review]
Looks good!

**Comment by kritibirda26** [comment]
@wenzeslaus Hi! Can you also review this PR so that it can be merged?

File: vector/v.univar/testsuite/v_univar_test.py
Comment (cwhite911) [inline vector/v.univar/testsuite/v_univar_test.py:1]: Rename test to `test_v_univar.py`

**Review by cwhite911 — CHANGES_REQUESTED** [review]

File: vector/v.univar/testsuite/v_univar_test.py
Comment (kritibirda26) [reply vector/v.univar/testsuite/v_univar_test.py:1]: Done

**Review by cwhite911 — APPROVED** [review]

**Review by echoix — APPROVED** [review]

**Pull request merged**